experimental/zerotrie/src/zerotrie.rs

```diff
@@ -0,0 +1,625 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::reader::*;
+
+use core::borrow::Borrow;
+
+#[cfg(feature = "alloc")]
+use crate::{builder::bytestr::ByteStr, builder::nonconst::ZeroTrieBuilder, error::Error};
+#[cfg(feature = "alloc")]
+use alloc::{boxed::Box, collections::BTreeMap, collections::VecDeque, string::String, vec::Vec};
+#[cfg(feature = "litemap")]
+use litemap::LiteMap;
+
+/// A data structure that compactly maps from byte sequences to integers.
+///
+/// There are several variants of `ZeroTrie` which are very similar but are optimized
+/// for different use cases:
+///
+/// - [`ZeroTrieSimpleAscii`] is the most compact structure. Very fast for small data.
+///   Only stores ASCII-encoded strings. Can be const-constructed!
+/// - [`ZeroTriePerfectHash`] is also compact, but it also supports arbitrary binary
+///   strings. It also scales better to large data. Cannot be const-constructed.
+/// - [`ZeroTrieExtendedCapacity`] can be used if more than 2^32 bytes are required.
+///
+/// You can create a `ZeroTrie` directly, in which case the most appropriate
+/// backing implementation will be chosen.
+///
+/// # Backing Store
+///
+/// The data structure has a flexible backing data store. The only requirement for most
+/// functionality is that it implement `AsRef<[u8]>`. All of the following are valid
+/// ZeroTrie types:
+///
+/// - `ZeroTrie<[u8]>` (dynamically sized type: must be stored in a reference or Box)
+/// - `ZeroTrie<&[u8]>` (borrows its data from a u8 buffer)
+/// - `ZeroTrie<Vec<u8>>` (fully owned data)
+/// - `ZeroTrie<ZeroVec<u8>>` (the recommended borrowed-or-owned signature)
```

Comment:
nit: I'd rather recommend the others, as they don't need the zerovec dependency

Reply:
I don't trust Serde with Cows (https://github.com/serde-rs/serde/issues/2016) and this crate is definitely in the zerovec ecosystem so I thought it was appropriate to plug zerovec as the recommended signature